src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java

```diff
@@ -1,62 +1,115 @@
+        } catch (Exception exception) {
+            log.error("{}:: Publishing message to service bus topic failed with exception: {}:: Job Id {}",
+                    loggingComponentName, exception, validationService.getAuditJobId());
+            if (Objects.nonNull(serviceBusSenderClient) && Objects.nonNull(transactionContext)) {
+                serviceBusSenderClient.rollbackTransaction(transactionContext);
+            }
+            throw new CaseworkerMessageFailedException(CaseWorkerConstants.ASB_PUBLISH_ERROR);
+        }
+        serviceBusSenderClient.commitTransaction(transactionContext);
         log.info("{}:: Message published to service bus topic:: Job Id is: {}", loggingComponentName,
                 validationService.getAuditJobId());
     }
 
-    @Recover
-    public void recoverMessage(Exception ex) {
-        log.error("{}:: Publishing message to service bus topic failed with exception: {}:: Job Id {}",
-                loggingComponentName, ex, validationService.getAuditJobId());
-        throw new CaseworkerMessageFailedException(CaseWorkerConstants.ASB_PUBLISH_ERROR);
+    private void publishMessageToTopic(PublishCaseWorkerData caseWorkerData,
+                                      ServiceBusSenderClient serviceBusSenderClient,
+                                      ServiceBusTransactionContext transactionContext) {
+        log.info("{}:: Started publishing to topic::", loggingComponentName);
+        ServiceBusMessageBatch messageBatch = serviceBusSenderClient.createMessageBatch();
+        List<ServiceBusMessage> serviceBusMessages = new ArrayList<>();
+
+        ListUtils.partition(caseWorkerData.getUserIds(), caseWorkerDataPerMessage)
+                .forEach(data -> {
+                    PublishCaseWorkerData publishCaseWorkerDataChunk = new PublishCaseWorkerData();
+                    publishCaseWorkerDataChunk.setUserIds(data);
+                    serviceBusMessages.add(new ServiceBusMessage(new Gson().toJson(publishCaseWorkerDataChunk)));
+                });
+
+        for (ServiceBusMessage message : serviceBusMessages) {
+            if (messageBatch.tryAddMessage(message)) {
+                continue;
+            }
+
+            // The batch is full, so we create a new batch and send the batch.
+            serviceBusSenderClient.sendMessages(messageBatch, transactionContext);
+
+            // create a new batch
+            messageBatch = serviceBusSenderClient.createMessageBatch();
+
+            // Add that message that we couldn't before.
+            if (!messageBatch.tryAddMessage(message)) {
+                log.error("Message is too large for an empty batch. Skipping. Max size: {}.",
+                        messageBatch.getMaxSizeInBytes());
+            }
+        }
+
+        if (messageBatch.getCount() > 0) {
+            serviceBusSenderClient.sendMessages(messageBatch, transactionContext);
+            log.info("Sent a batch of messages to the topic: {}", topic);
```

Comment:
Can we add loggingComponentName  and job id as well for tracking purpose ?

Reply:
Makes sense. Adding.